app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,157 @@
-    info "Starting DeleteObjectHierarchyWorker##{method_name} with the hierarchy of workers: #{workers_hierarchy}"
-    object = GlobalID::Locator.locate(options['object_global_id'])
-    background_destroy_method = @background_destroy_method.presence || 'destroy'
-    DeletePlainObjectWorker.perform_later(object, workers_hierarchy, background_destroy_method)
-    info "Finished DeleteObjectHierarchyWorker##{method_name} with the hierarchy of workers: #{workers_hierarchy}"
-  rescue ActiveRecord::RecordNotFound => exception
-    info "DeleteObjectHierarchyWorker##{method_name} raised #{exception.class} with message #{exception.message}"
+  after_perform do |job|
+    info "Finished #{job.class}#perform with the hierarchy of workers: #{job.arguments}"
   end
 
-  protected
-
-  delegate :info, to: 'Rails.logger'
+  class << self
+    # convenience method to schedule deleting an active record object
+    def delete_later(ar_object)
+      perform_later *hierarchy_entries_for(ar_object)
+    end
 
-  attr_reader :object, :caller_worker_hierarchy
+    private
 
-  def build_batch
-    batch = Sidekiq::Batch.new
-    batch.description = batch_description
-    batch_callbacks(batch) { batch.jobs { destroy_and_delete_associations } }
-    batch
-  end
+    # @return the hierarchy entries to handle deletion of an object
+    def hierarchy_entries_for(ar_object)
+      return [] unless ar_object.persisted? # e.g. when calling Proxy#oidc_configuration a new object can be generated
 
-  def batch_description
-    "Deleting #{object.class.name} [##{object.id}]"
-  end
+      if ar_object.is_a?(Account) && !ar_object.should_be_deleted?
+        raise DoNotRetryError, "background deleting account #{ar_object.id} which is not scheduled for deletion"
+      end
 
-  def batch_callbacks(batch)
-    %i[success complete].each { |name| batch.on(name, self.class, callback_options) }
-    yield
-    bid = batch.bid
+      ar_object_str = "#{ar_object.class}-#{ar_object.id}"
 
-    if Sidekiq::Batch::Status.new(bid).total.zero?
-      on_complete(bid, callback_options)
-    else
-      info("DeleteObjectHierarchyWorker#batch_success_callback retry job with the hierarchy of workers: #{caller_worker_hierarchy}")
-      retry_job wait: 5.minutes
+      [
+        "Plain-#{ar_object_str}",
+        *ar_object.background_deletion.map { "Association-#{ar_object_str}:#{_1}" },
+      ]
     end
   end
 
-  def destroy_and_delete_associations
-    Array(object.background_deletion).each do |association_config|
-      reflection = BackgroundDeletion::Reflection.new(association_config)
-      next unless destroyable_association?(reflection.name)
+  # @param hierarchy [Array<String>] something like ["Plain-Service-1234", "Association-Service-1234:plans" ...]
+  # @note processes deleting a hierarchy of objects and reschedules itself at current progress after a time limit
+  def perform(*hierarchy)
+    return compatibility(hierarchy) unless hierarchy.first.is_a?(String)
```

Comment:
What is this method for? apparently the `hierarchy` parameter, called `_object` inside the method, is just being ignored.

Reply:
Correct. I don't care about the object. The idea is to take the root object of the original operation and schedule it anew with the new worker logic. The job locking should take care of discarding the pile of duplicates that could be present with the old logic.

Reply:
Thanks @mayorova. So this is how `hierarchy` was before this PR. But in that case, is `"gid://system/User/17"` not a `String`? Because the condition to call `compatibility` is `unless hierarchy.first.is_a?(String)`

Reply:
OK, thanks for clarifying. One more question: in the case we have a job which was enqueued with only an object as param, like from https://github.com/3scale/porta/blob/master/app/models/backend_api.rb#L109, then the `compatibility` will do nothing and return, as I understand it:

https://github.com/3scale/porta/blob/3612c2fa9db9662be2c4894c559ec49812407f9d/app/workers/delete_object_hierarchy_worker.rb#L135-L138

Then we are losing that job, or am I wrong?

Reply:
Oh, yes, I had the same question actually.
If the job has just been scheduled (with the old code), but not processed yet (so, no hierarchy has not been populater), I think we need to still process it, based on the first argument (object).

Reply:
Good catch. Although this should be a super low chance.

Reply:
Should be fixed with https://github.com/3scale/porta/pull/3958/commits/4395034e8a4e17aa3990258cdbb86e0c2ffd79cb